### ISSUE TITLE
cannot save gplearn model

### ISSUE THREAD

**Issue by Alsac** [trevorstephens/gplearn#161]
When I use pickle.dump to save fitted model with custom function, pickle go wrong as like:_pickle.PicklingError:can't pickle <function _sigmoid at 0x7facce5dcd90>:it's not the same object as __main__.sigmoid.
I find dump model without custom function is work.
How could I save model with custom function?Thx

**Comment by Alsac** [comment]
well ,I solve it. because I made function_set as tuple. when I use list replace, it works.

**Comment by ElvenMonky** [comment]
This solution doesn't work for me.

**Comment by trevorstephens** [comment]
I can't help if no one tells me what the problem is @ElvenMonky . Open a new issue, choose bug, and follow the instructions please.

**Comment by trevorstephens** [comment]
`function_set` should accept a tuple per the doc string. will investigate @Alsac ... Might need to specify list is required if this is (strangely) a pickle problem.

**Comment by trevorstephens** [comment]
Some kind of short example that can replicate the problem would be extremely helpful

**Comment by ElvenMonky** [comment]
If Kaggle public kernel could work as example, I'll prepare one ASAP.

**Comment by trevorstephens** [comment]
Sure, scikit-learn's toy datasets even better 👍 

    from sklearn.datasets import make_blobs
    X, y = make_blobs(random_state=0, n_features=4, centers=2)

or something like that :-)

**Comment by trevorstephens** [comment]
Doesn't sound like a data-dependent issue so generic data should be able to illustrate the problem hopefully, and make it easier to replicate outside of Kaggle's environment

**Comment by Alsac** [comment]
It's extremely strange that it go wrong again. last time I replaced funcset type from tuple to list,and it works many days. recently it go wrong. And I find when n_jobs equals to 1 is work(can save with pickle), equals others is not(with error as describe before). I guess the problem is related to parallel. so far I haven't find solution for that.🤖

**Comment by trevorstephens** [comment]
```
import numpy as np
import pickle
from sklearn.datasets import load_boston
from gplearn.functions import make_function
from gplearn.genetic import SymbolicRegressor

boston = load_boston()

def _logical(x1, x2, x3, x4):
    return np.where(x1 > x2, x3, x4)

logical = make_function(function=_logical,
                        name='logical',
                        arity=4)

est = SymbolicRegressor(generations=5,
                        function_set=('add', 'sub', 'mul', 'div', logical),
                        n_jobs=-1)
est.fit(boston.data, boston.target)
pickle_object = pickle.dumps(est)
```

Fails with `n_jobs=-1`, works with `n_jobs=1` on my Linux environment running python 3.6 and scikit-learn master branch.

Looks to be related to the parallel process. A different `joblib` backend might solve this but could require upgrade of dependencies. Might also help out on Windows environments like #71 . Shall investigate.

For now, run single-threaded while I work on a fix.

**Comment by trevorstephens** [comment]
@Alsac @ElvenMonky I think I have found a fix.

`pip install --upgrade joblib`

This should get you to 0.13.2... I can refine versions later but looking for some feedback now before any changes to master if possible. I think decorating your function before calling the `make_function` factory will solve this (I hope!)

```
from joblib import wrap_non_picklable_objects

@wrap_non_picklable_objects
def _my_custom_function(x1, x2):
    ...
    return things

my_fun = make_function(function=_my_custom_function,
                        name='my_custom_function',
                        arity=2)

est = SymbolicRegressor(...)
est.fit(...)
pickle_object = pickle.dumps(est)
```

Can either, or both, of you give this a crack and let me know if it works as expected on your systems?

**Comment by Alsac** [comment]
Sounds great! I will try this method tomorrow and give you feedback.

**Comment by trevorstephens** [comment]
Thanks! Don't forget to try n_jobs > 1 for multiprocessing which it seems is where this bug lies.

**Comment by Alsac** [comment]
That is great！It solve the problem perfectly. Currently I tried decorating function and it works（dump model success） when n_jobs=-1, My environment is python3.6 in Linux😃

**Comment by trevorstephens** [comment]
Great news, thanks for letting me know @Alsac 😃 I'll try to investigate some ways to make this invisible to the user over the next week or so. But for now it looks like we have a workable solution that could be solved by documentation at the worst case.

**Comment by ElvenMonky** [comment]
@trevorstephens Thanks, your solution solved the problem for me as well! Was hard to guess that it's `n_jobs=-1` that actually causes a problem.

**Comment by trevorstephens** [comment]
Great, thanks @ElvenMonky :smiley: I should have a more user-friendly fix in soon. Plan is to just do the function decoration for you at run-time if you are running parallel.

**Comment by trevorstephens** [comment]
@ElvenMonky @Alsac I have made a new bugfix release available now on PyPI for installation using pip. If you have the opportunity to try it out it would be a great help. Feel free to raise a new issue should you run into any trouble. :+1:

**Comment by Alsac** [comment]
thank u👍
